src/qttools/datastructures/dsbcoo.py

```diff
@@ -231,7 +231,7 @@ def __iadd__(self, other: "DSBSparse | sparse.sparray") -> "DSBCOO":
         """In-place addition of two DSBSparse matrices."""
         if sparse.issparse(other):
             lil = other.tolil()
-            sparray_data = lil[self.rows, self.cols]
+            sparray_data = lil[self.rows, self.cols].toarray()
```

Comment:
Does the in-place addition/substraction of 2 DSBSparse matrices densify the blocks being touched? Does it retain/modify the sparsity pattern of the blocks or assume that the sparsity patterns of the 2 matrices matches?

Reply:
To clarify: This change concerns the addition of a `scipy.sparse.sparray` to the `DSBCOO`. This is used to construct the bare system matrices from the hamiltonian `sparray`, etc.

The sparsity of the `DSBCOO` is assumed to always be fixed, so here we also only extract the elements from the `sparray` that match, i.e. `(self.rows, self.cols)`. In `scipy` only the `lil_array` supports this fancy-indexing, but it returns a new `lil_array` (of shape `(1, len(self.rows))`), thus the call to `.toarray()`.

I will add a test for this over on the datastructures branch, it's not covered atm :cowboy_hat_face:.

The in-place operations of two `DSBCOO` matrices always operates on the data buffers directly, so nothing gets densified and sparsity pattern remain unchanged. For this the two matrices have to be commensurable:
https://github.com/vincent-maillou/qttools/blob/609f8ddc6ba061da3b0b748a374d77eed7acf720/src/qttools/datastructures/dsbcoo.py#L213-L228

Is that alright from your side?

Reply:
Yep sounds good! thanks for the explanations